File: matrix.py
#!/usr/bin/python3
import argparse
import fileinput
import sys


class Matrix(object):

  def __init__(self):
    self.rows = []
    self.cols = []
    self.body = {}

  def add(self, r, c, b):
    if r not in self.rows:
      self.rows.append(r)
    if c not in self.cols:
      self.cols.append(c)
    self.body[r, c] = '' if b is None else b

  def print(self, file, sep, missing):
    rows, cols = self.rows, self.cols
    print('', *cols, sep=sep, file=file)

    for r in rows:
      cells = [self.body.get((r, c), missing) for c in cols]
      print(r, *cells, sep=sep, file=file)


if __name__ == '__main__':
  parser = argparse.ArgumentParser()
  parser.add_argument('files', nargs='*')
  parser.add_argument('--input-sep')
  parser.add_argument('--output-sep', default='\t')
  parser.add_argument('--missing', default='')
  args = parser.parse_args()

  m = Matrix()
  for line in fileinput.input(args.files):
    line = line.rstrip()
    items = line.split(args.input_sep, maxsplit=2)
    if len(items) <= 1:
      r, = items
      c = ''
      b = None
    elif len(items) <= 2:
      r, c = items
      b = None
    else:
      r, c, b = items
    m.add(r, c, b)
  m.print(sys.stdout, sep=args.output_sep, missing=args.missing)

File: matrix.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
)

type key struct{ Row, Col string }

type matrix struct {
	rows    []string
	rowsMap map[string]bool
	cols    []string
	colsMap map[string]bool
	body    map[key]string
}

func newMatrix() *matrix {
	return &matrix{
		rowsMap: make(map[string]bool),
		colsMap: make(map[string]bool),
		body:    make(map[key]string),
	}
}

func (m *matrix) Add(r, c, b string) {
	if !m.rowsMap[r] {
		m.rows = append(m.rows, r)
		m.rowsMap[r] = true
	}
	if !m.colsMap[c] {
		m.cols = append(m.cols, c)
		m.colsMap[c] = true
	}
	m.body[key{Row: r, Col: c}] = b
}

func (m *matrix) Print(w io.Writer, output_sep, missing string) {
	for _, c := range m.cols {
		fmt.Fprint(w, output_sep, c)
	}
	fmt.Fprintln(w)
	for _, r := range m.rows {
		fmt.Fprint(w, r)
		for _, c := range m.cols {
			b, ok := m.body[key{Row: r, Col: c}]
			if !ok {
				b = missing
			}
			fmt.Fprint(w, output_sep, b)
		}
		fmt.Fprintln(w)
	}
}

var (
	inputSep  = flag.String("input-sep", "\t", "input separator")
	outputSep = flag.String("output-sep", "\t", "output separator")
	missing   = flag.String("missing", "", "missing value indicator")
)

func main() {
	flag.Parse()

	m := newMatrix()
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		items := strings.SplitN(line, *inputSep, 3)
		var r, c, b string
		switch {
		case len(items) <= 1:
			r = items[0]
		case len(items) <= 2:
			r = items[0]
			c = items[1]
		default:
			r = items[0]
			c = items[1]
			b = items[2]
		}
		m.Add(r, c, b)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	m.Print(os.Stdout, *outputSep, *missing)
}
